### ISSUE TITLE
[feature request] Gumbel Softmax for 3D tensor

### ISSUE THREAD

**Issue by shinochin** [pytorch/pytorch#12643]
## 🚀 Feature

I want F.gumbel_softmax to support not only 2D tensors but also 3D or 4D tensors.

## Motivation

I want to apply Gumbel Softmax to 3D tensors.

## Additional context

I think it is easy to make Gumbel Softmax available for Tensors of any number of dimensions.
Are there any reasons why you haven't done yet?

cf.
https://github.com/pytorch/pytorch/blob/a98958d3bdafd4d20e53479b7b070127628a2249/torch/nn/functional.py#L1043
